### PR TITLE
[std.meta] Remove workarounds for an alias parameter not accepting basic types

### PR DESCRIPTION
**Pull request by ntrel**
Alias parameters accept basic types so various overloads are unnecessary now. E.g.:
```d
template Replace(T, U, TList...);
template Replace(alias T, U, TList...);
template Replace(T, alias U, TList...);
template Replace(alias T, alias U, TList...);
```
Only the last one is needed.

This also cleans up e.g. `GenericReplace` which the overloads forward to, and for those I've renamed `TList` to `Seq`. I've also replaced some usage of `OldAlias` with `Alias`.

### CODE REVIEW

**Comment by ntrel** [comment]
> generated/linux/debug/64/publictests/std_traits.d(5111): Error: static assert:  `is((Object, A, B) == (const(Object), const(A), const(B)))` is false

This failure is due to:
>         alias ImplicitConversionTargets = staticMap!(ApplyLeft!(CopyConstness, T), TransitiveBaseTypeTuple!(T));

`ApplyLeft!(CopyConstness, T)` is failing, as `Alias!(const C)` now gives `C` (not `const C`), due to:
https://issues.dlang.org/show_bug.cgi?id=20863

File: std/meta.d [hunk 1]
```diff
@@ -267,12 +267,23 @@ if (!isAggregateType!T || is(Unqual!T == T))
...
+template staticIndexOf(alias T, Seq...)
 {
-    enum staticIndexOf = genericIndexOf!(T, TList);
+    static foreach (idx, arg; Seq)
+    {
+        static if (is(typeof(staticIndexOf) == void) && // not yet defined
+                   isSame!(T, arg))
+        {
+            enum staticIndexOf = idx;
+        }
```
Comment (MoonlightSentinel) [inline std/meta.d:281]: Simply `break` out of the loop?

```suggestion
    Loop:
    static foreach (idx, arg; Seq)
    {
        static if (isSame!(T, arg))
        {
            enum staticIndexOf = idx;
            break Loop;
        }
```

Comment (ntrel) [reply std/meta.d:281]: @MoonlightSentinel:
From https://dlang.org/spec/version.html#break-continue:
> Note that static foreach itself cannot be broken nor continued even if it is explicitly labeled

(I haven't actually tried it recently though). Also I didn't write the code, the diff just makes it look like that. I'm just renaming some identifiers ;-)

Comment (MoonlightSentinel) [reply std/meta.d:281]: You're right, I misremembered the interaction between `static foreach` and `break`.

Comment (MoonlightSentinel) [reply std/meta.d:281]: Im curious how this implementation compares to something like this:

```
template staticIndexOf(alias T, Seq...)
{
    enum staticIndexOf = ()
    {
        foreach(idx, arg; Seq)
        {
            if (isSame!(T, arg))
                return idx;
        }
        
        return -1;
    }();
}
```

Comment (ntrel) [reply std/meta.d:281]: Interesting solution. I don't know, but for large sequence length I think static foreach is v slow, at least ATM:
http://dpldocs.info/this-week-in-d/Blog.Posted_2020_05_11.html#dmd-and-static-foreach

Comment (ntrel) [reply std/meta.d:281]: That instantiates `isSame` for every element of `Seq`, which could be expensive if there's a match near the start.

**Review by andralex — APPROVED** [review]
love the red

**Comment by RazvanN7** [comment]
@ntrel What is the status of this PR?

**Comment by ntrel** [comment]
@RazvanN7 this depends on https://github.com/dlang/dmd/pull/11320/

**Comment by LightBender** [comment]
@jmdavis does this still apply to Phobos 3 traits?

**Pull request closed without merging**